File: container/v0/traffic.py
from bcc import BPF
from time import sleep
from BaseHTTPServer import BaseHTTPRequestHandler, HTTPServer
import socket
import os
import sys
import json

L4TcpSendV4 = 0
L4TcpRecvV4 = 1
L4UdpSendV4 = 2
L4UdpRecvV4 = 3
L3TcpSendV4 = 4
L3UdpSendV4 = 5
L3RawSendV4 = 6
L4TcpSendV6 = 7
L4TcpRecvV6 = 8
L4UdpSendV6 = 9
L4UdpRecvV6 = 10
L3TcpSendV6 = 11
L3UdpSendV6 = 12
L3RawSendV6 = 13

if len(sys.argv) != 2:
    print("Usage: %s <SOCK_FILE>" % (sys.argv[0]))
    exit(1)

SOCK_FILE = sys.argv[1]

class UnixSocketRequestHandler(BaseHTTPRequestHandler):
    def load_bpf(self):
        try:
            print("load bpf program")
            b = BPF(src_file="./data/traffic.c")

            print("attach kprobe and kretprobe")
            b.attach_kprobe(event="udp_sendmsg", fn_name="kprobe__udp_sendmsg")
            b.attach_kretprobe(event="udp_sendmsg", fn_name="kretprobe__udp_sendmsg")
            b.attach_kprobe(event="udpv6_sendmsg", fn_name="kprobe__udpv6_sendmsg")
            b.attach_kretprobe(event="udpv6_sendmsg", fn_name="kretprobe__udpv6_sendmsg")
            b.attach_kretprobe(event="skb_consume_udp", fn_name="kprobe__skb_consume_udp")
            b.attach_kprobe(event="tcp_sendmsg", fn_name="kprobe__tcp_sendmsg")
            b.attach_kretprobe(event="tcp_sendmsg", fn_name="kretprobe__tcp_sendmsg")
            b.attach_kprobe(event="tcp_cleanup_rbuf", fn_name="kprobe__tcp_cleanup_rbuf")
            b.attach_kprobe(event="inet_csk_destroy_sock", fn_name="kprobe__inet_csk_destroy_sock")
            b.attach_kprobe(event="ip_output", fn_name="kprobe__ip_output")
            b.attach_kprobe(event="ip6_output", fn_name="kprobe__ip6_output")

            print("load bpf program success")
            UnixSocketRequestHandler.bpf = b

            self.send_response(200)
            self.send_header('Content-Type', 'text/plain')
            self.end_headers()
            self.wfile.write("success")
        except Exception as e:
            self.send_response(500)
            self.send_header('Content-Type', 'text/plain')
            self.end_headers()
            self.wfile.write(str(e))

    def collect(self, flows, name, v4i, v6i):
        for k, v in self.bpf[name].items():
            if k.container_id == "":
                cid = "000000000000"
            else:
                cid = k.container_id
            bytes = sum(v)

            flow = flows.get(cid)
            if flow == None:
                flow = [0 for i in range(14)]
                flows[cid] = flow
            if k.family == 2:
                flow[v4i] += bytes
            else:
                flow[v6i] += bytes

    def get_flow(self):
        try:
            flows = {}
            self.collect(flows, "L4_tcp_send_bytes", L4TcpSendV4, L4TcpSendV6)
            self.collect(flows, "L4_tcp_recv_bytes", L4TcpRecvV4, L4TcpRecvV6)
            self.collect(flows, "L4_udp_send_bytes", L4UdpSendV4, L4UdpSendV6)
            self.collect(flows, "L4_udp_recv_bytes", L4UdpRecvV4, L4UdpRecvV6)
            self.collect(flows, "L3_tcp_send_bytes", L3TcpSendV4, L3TcpSendV6)
            self.collect(flows, "L3_udp_send_bytes", L3UdpSendV4, L3UdpSendV6)
            self.collect(flows, "L3_raw_send_bytes", L3RawSendV4, L3RawSendV6)
            body = json.dumps(flows)
            self.send_response(200)
            self.send_header('Content-Type', 'application/json')
            self.end_headers()
            self.wfile.write(body)
        except Exception as e:
            self.send_response(500)
            self.send_header('Content-Type', 'text/plain')
            self.end_headers()
            self.wfile.write(str(e))

    def do_GET(self):
        if self.path == "/load":
            self.load_bpf()
        else:
            self.get_flow()

    def log_message(self, format, *args):
        pass

class UnixSocketHTTPServer(HTTPServer):
    address_family = socket.AF_UNIX

    def server_bind(self):
        if os.path.exists(self.server_address):
            os.unlink(self.server_address)
        HTTPServer.server_bind(self)

try:
    print("run http server")
    httpd = UnixSocketHTTPServer(SOCK_FILE, UnixSocketRequestHandler)
    httpd.serve_forever()
except KeyboardInterrupt:
    print("stop server")
    exit(0)
except Exception as e:
    print("error: %s" % str(e))
    exit(1)

File: container/v2/traffic.go
package traffic

import (
	"encoding/binary"
	"go_minion/internal/collector/types"
	"go_minion/pkg/logger"
	"go_minion/pkg/utility"

	"github.com/cilium/ebpf"
	"github.com/cilium/ebpf/link"
)

//go:generate go run github.com/cilium/ebpf/cmd/bpf2go -target amd64 traffic traffic.c
//go:generate go run github.com/cilium/ebpf/cmd/bpf2go -target arm64 traffic traffic.c
//go:generate go run github.com/cilium/ebpf/cmd/bpf2go -target arm traffic traffic.c

type Collector struct {
	logger *logger.FieldLogger
	objs   trafficObjects
	links  map[string]link.Link
}

func NewCollector(fileLogger *logger.FileLogger) (*Collector, error) {
	c := &Collector{
		logger: logger.NewFieldLogger(fileLogger, "ebpf-container-v2"),
		links:  make(map[string]link.Link),
	}
	if err := c.load(); err != nil {
		c.Close()
		return nil, err
	}
	return c, nil
}

func (c *Collector) load() error {
	if err := loadTrafficObjects(&c.objs, nil); err != nil {
		return err
	}

	if link, err := link.AttachTracing(link.TracingOptions{Program: c.objs.FexitTcpSendpage}); err != nil {
		return err
	} else {
		c.links["fexit/tcp_sendpage"] = link
	}

	if link, err := link.AttachTracing(link.TracingOptions{Program: c.objs.FexitTcpSendmsg}); err != nil {
		return err
	} else {
		c.links["fexit/tcp_sendmsg"] = link
	}

	if link, err := link.Kprobe("tcp_cleanup_rbuf", c.objs.KprobeTcpCleanupRbuf, nil); err != nil {
		return err
	} else {
		c.links["kprobe/tcp_cleanup_rbuf"] = link
	}

	if link, err := link.Kprobe("inet_csk_destroy_sock", c.objs.KprobeInetCskDestroySock, nil); err != nil {
		return err
	} else {
		c.links["kprobe/inet_csk_destroy_sock"] = link
	}

	if link, err := link.AttachTracing(link.TracingOptions{Program: c.objs.FexitUdpSendmsg}); err != nil {
		return err
	} else {
		c.links["fexit/udp_sendmsg"] = link
	}

	if link, err := link.AttachTracing(link.TracingOptions{Program: c.objs.FexitUdpv6Sendmsg}); err != nil {
		return err
	} else {
		c.links["fexit/udpv6_sendmsg"] = link
	}

	if link, err := link.Kprobe("skb_consume_udp", c.objs.KprobeSkbConsumeUdp, nil); err != nil {
		return err
	} else {
		c.links["kprobe/skb_consume_udp"] = link
	}

	if link, err := link.Kprobe("ip_output", c.objs.KprobeIpOutput, nil); err != nil {
		return err
	} else {
		c.links["kprobe/ip_output"] = link
	}

	if link, err := link.Kprobe("ip6_output", c.objs.KprobeIp6Output, nil); err != nil {
		return err
	} else {
		c.links["kprobe/ip6_output"] = link
	}

	return nil
}

func (c *Collector) Close() {
	for _, link := range c.links {
		link.Close()
	}
	c.objs.Close()
}

func (c *Collector) CollectFlows(flows types.FlowMap, name string, m *ebpf.Map, v4Index int, v6Index int) {
	cpus := utility.GetNumOfPossibleCpus()
	key := make([]byte, 16)
	values := make([]uint64, cpus)
	for i := m.Iterate(); i.Next(&key, &values); {
		family := binary.LittleEndian.Uint16(key[0:2])
		container := string(key[2:14])
		bytes := utility.Sum(values)

		// 将空容器ID转为可打印字符
		if container == "\000\000\000\000\000\000\000\000\000\000\000\000" {
			container = "000000000000"
		}

		switch family {
		case types.AF_INET:
			c.logger.Debug("name=%s proto=IPv4 container=%s bytes=%d", name, container, bytes)
			v, ok := flows[container]
			if ok {
				v[v4Index] += bytes
			} else {
				v = &types.FlowData{}
				v[v4Index] = bytes
				flows[container] = v
			}
		case types.AF_INET6:
			c.logger.Debug("name=%s proto=IPv6 container=%s bytes=%d", name, container, bytes)
			v, ok := flows[container]
			if ok {
				v[v6Index] += bytes
			} else {
				v = &types.FlowData{}
				v[v6Index] = bytes
				flows[container] = v
			}
		}
	}
}

func (c *Collector) Collect() types.FlowMap {
	flows := make(map[string]*types.FlowData)
	c.CollectFlows(flows, "L4TcpSend", c.objs.L4TcpSendBytes, int(types.L4TcpSendV4), int(types.L4TcpSendV6))
	c.CollectFlows(flows, "L4TcpRecv", c.objs.L4TcpRecvBytes, int(types.L4TcpRecvV4), int(types.L4TcpRecvV6))
	c.CollectFlows(flows, "L4UdpSend", c.objs.L4UdpSendBytes, int(types.L4UdpSendV4), int(types.L4UdpSendV6))
	c.CollectFlows(flows, "L4UdpRecv", c.objs.L4UdpRecvBytes, int(types.L4UdpRecvV4), int(types.L4UdpRecvV6))
	c.CollectFlows(flows, "L3TcpSend", c.objs.L3TcpSendBytes, int(types.L3TcpSendV4), int(types.L3TcpSendV6))
	c.CollectFlows(flows, "L3UdpSend", c.objs.L3UdpSendBytes, int(types.L3UdpSendV4), int(types.L3UdpSendV6))
	c.CollectFlows(flows, "L3RawSend", c.objs.L3RawSendBytes, int(types.L3RawSendV4), int(types.L3RawSendV6))
	return flows
}
